### PR TITLE
Allow to retry failed test from test context for #1773

### PR DESCRIPTION
**Pull request by longlho**
- make retries run proper hooks
- allow retries override at different levels

[Review on Reviewable](https://reviewable.io/reviews/mochajs/mocha/1989)

### CODE REVIEW

File: lib/suite.js [hunk 1]
```diff
@@ -108,6 +110,25 @@ Suite.prototype.timeout = function(ms) {
 };
 
 /**
+ * Set timeout `ms` or short-hand such as "2s".
```
Comment (danielstjules) [inline lib/suite.js]: Doc comments need updating?

File: test/acceptance/retries.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
+describe('retries', function() {
```
Comment (danielstjules) [inline test/acceptance/retries.js]: Could this be moved to test/integration as a test fixture? Makes it easier to verify test output

**Comment by danielstjules** [comment]
Looks like you may have accidentally deleted mocha.js as part of that commit? Could you amend and restore it?

**Comment by longlho** [comment]
@danielstjules my bad :) commit ammended

File: lib/suite.js (hunk 1 above)
Comment (longlho) [inline lib/suite.js]: :+1:

File: test/acceptance/retries.js (hunk 2 above)
Comment (longlho) [inline test/acceptance/retries.js]: yup moved :)

File: lib/mocha.js [hunk 3]
```diff
@@ -373,6 +377,20 @@ Mocha.prototype.timeout = function(timeout) {
...
 /**
+ * Set the number of times to retry failed tests.
+ *
+ * @param {Number} retry times
+ * @return {Mocha}
+ * @api public
+ * @param {number} retry times
+ * @return {Mocha}
+ */
+Mocha.prototype.retires = function(retries) {
```
Comment (tinganho) [inline lib/mocha.js]: Typo here `retires` -> `retries`.

File: lib/runner.js [hunk 4]
```diff
@@ -529,8 +529,12 @@ Runner.prototype.runTests = function(suite, fn) {
         test = self.test;
 
         if (err) {
+          var retry = test.currentRetry();
           if (err instanceof Pending) {
             self.emit('pending', test);
+          } else if (retry < test.retries()) {
+            test.currentRetry(retry + 1);
+            tests.unshift(test);
```
Comment (tinganho) [inline lib/runner.js]: Do we want to add more tests to the test suite?

File: lib/suite.js [hunk 5]
```diff
@@ -108,6 +110,25 @@ Suite.prototype.timeout = function(ms) {
...
+ * @api private
+ * @param {number|string} n
+ * @return {Suite|number} for chaining
+ */
+Suite.prototype.retries = function(n) {
+  if (!arguments.length) {
+    return this._retries;
+  }
+  if (n.toString() === '0') {
+    this._retries = -1;
```
Comment (tinganho) [inline lib/suite.js]: Returning `-1` here when the retries are `0` I think is very error prone.

File: lib/mocha.js (hunk 3 above)
Comment (tinganho) [inline lib/mocha.js]: Also the parameter name should be `n`? At least it is on your other methods.

Comment (longlho) [inline lib/mocha.js]: ah my bad, fixed

File: lib/suite.js (hunk 5 above)
Comment (longlho) [inline lib/suite.js]: hmm doesn't really matter IMO, as long as `_currentRetry > _retries`. It's more like an indicator.

File: lib/runner.js (hunk 4 above)
Comment (longlho) [inline lib/runner.js]: up to the collaborators I guess.

**Comment by longlho** [comment]
ping @boneskull @danielstjules

File: lib/runner.js (hunk 4 above)
Comment (danielstjules) [inline lib/runner.js]: Given `var tests = suite.tests.slice();`, adding it to `tests` is a fine approach. As long as `suite.tests` isn't being modified :)

File: lib/suite.js (hunk 5 above)
Comment (danielstjules) [inline lib/suite.js]: Seems like `if (n.toString() === '0') this._retries = -1;` is probably unnecessary? If a user did `.retries(-1)`, `.retries(0)`, `.retries(false)` etc, it'd be falsey anyway, as long as you make sure `_currentRetry` starts at 1

Comment (danielstjules) [inline lib/suite.js]: Woops, correction. _currentRetry starting at 0 is fine

File: test/integration/fixtures/retries/early-pass.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+describe('retries', function() {
+  this.retries(4);
```
Comment (danielstjules) [inline test/integration/fixtures/retries/early-pass.js]: `this.retries(1)` probably makes sense here?

**Comment by danielstjules** [comment]
Great job on the tests :)

File: lib/suite.js (hunk 5 above)
Comment (longlho) [inline lib/suite.js]: yeah I'll just simplify this to take in `n`

**Comment by longlho** [comment]
@danielstjules tks :) changes made

**Comment by danielstjules** [comment]
Thanks! And can you `git checkout 67d0facb5153feb7db4d8e161778436a75117189 mocha.js && git commit --amend` Not sure why it's got a diff :)

**Comment by longlho** [comment]
done :)

**Comment by danielstjules** [comment]
Playing around with this branch today and will let you know if I have anymore feedback :)

File: lib/mocha.js [hunk 7]
```diff
@@ -373,6 +377,20 @@ Mocha.prototype.timeout = function(timeout) {
 };
 
 /**
+ * Set the number of times to retry failed tests.
+ *
+ * @param {Number} retry times
+ * @return {Mocha}
```
Comment (danielstjules) [inline lib/mocha.js]: Duplicate annotations here

File: test/integration/fixtures/retries/nested.js [hunk 8]
```diff
@@ -0,0 +1,9 @@
+describe('retries', function() {
+  this.retries(3);
+  describe('nested', function () {
+    it('should retry on test 3', function(){
```
Comment (danielstjules) [inline test/integration/fixtures/retries/nested.js]: The name of this test is a little ambiguous. This test will have only a single retry, and will fail on both attempts

File: test/integration/fixtures/retries/early-pass.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
+describe('retries', function() {
+  this.retries(1);
+  var times = 0;
+
+  it('should quit early', function() {
```
Comment (danielstjules) [inline test/integration/fixtures/retries/early-pass.js]: Maybe call it "should pass after 1 retry"?

**Comment by danielstjules** [comment]
Comments aside, the feature looks good to me. :) It behaves exactly as I'd expect:
- afterEach hooks are ran after test failures, before a subsequent retry
- Despite multiple retries, tests are only emitted once to reporters
- `.retries()` precedence makes sense: cli < suite < test. As an example, `this.retries(3);` within a test body will take precedence over `this.retries(1);` at the suite level, or any `--retries n` cli arg

Integration tests validate that hook behavior is correct with retries. Unless @mochajs/mocha has any objections, I hope to merge this after the comments are fixed. I think it'll fit nicely with a new minor release! :)

**Comment by tinganho** [comment]
I think mutating the `tests` variable in [here](https://github.com/mochajs/mocha/pull/1989/files#r46524902) is quite bad. Just think about it, an outside developer now need to learn this special rule. Although, IMHO this is an "ugly hack"  to support retries.

**Comment by danielstjules** [comment]
> Just think about it, an outside developer now need to learn this special rule.

What special rule? A developer just needs to read `runTests`, and it's quite clear: `var tests = suite.tests.slice();` https://github.com/longlho/mocha/blob/1773-retries/lib/runner.js#L436 It should be quite evident that it's a shallow copy of the suite.tests array :)

File: lib/mocha.js (hunk 7 above)
Comment (longlho) [inline lib/mocha.js]: ah, I fixed that, but turns out there're actually a bunch of other methods w/ dupe annotations (`timeout`, `enableTimeout`...)

**Comment by longlho** [comment]
@danielstjules changes made :)

**Comment by danielstjules** [comment]
Awesome! Thanks again for all your work on this! I'm sure a lot of people are looking forward to using it :) Should be pretty helpful for browser-based acceptance test suites.

**Comment by tinganho** [comment]
Does this support retrying beforeEach and afterEach hooks too?

I have tests that does HTTP calls in `beforeEach` hook and `afterEach` hook and they stall sometimes.

**Comment by tinganho** [comment]
Also maybe try to retry `after` and `before` hooks as well. I have some HTTP calls there too that have got stalled.

**Comment by danielstjules** [comment]
> Does this support retrying beforeEach and afterEach hooks too?

No, just tests, and I'd personally be opposed to the idea of retries for hooks. Hooks serve as setup/teardowns for tests, and can thus be used to cleanup before/after a failed test. For failed hooks (afterEach, after), you can't be sure of the state when retrying. If you have brittle logic in hooks, I'd suggest making a helper function and moving it to be invoked from the specs.

**Comment by ajaykodali** [comment]
Awesome! Just saw that this work has been merged and released. Been looking forward to this one. Thanks guys.

**Comment by dmitriiabramov** [comment]
regarding retries in before and after hooks.
would an API like this work?

```js
afterEach(function() {
  if(this.isThereAnyErrorsInTheConsole()) {
    this.retry(); // fails the test and retries it again
  }
});
```

that's just an explicit retry call that doesn't happen automatically.

in my project, we check for react `propTypes` and other stuff validation errors after the test passed.
When running e2e tests it becomes very flaky, mostly because of some random errors (like an image didn't load, or some third party ads library freaked out and threw up in the console)
the current options are:
1. Use an external helper and call it after every test (cons: people will forget to do it and there is not much you can do about it)
2. Create a global wrapper `global.it = wrapInConsoleOutputValidator(global.it)` (cons: hacky as hell)

this is also related to this issue, i suppose... https://github.com/mochajs/mocha/issues/2120

**Comment by kashinross** [comment]
I'm currently using this on some of my tests and I love the functionality. I'm currently using `this.retries(1)` at the describe level but I noticed the retry logic always attempts to run the current `it()`. Is there any way to retry the entire describe block on a failure?

**Pull request merged**